index.d.ts

```diff
@@ -0,0 +1,758 @@
+   * @param {Array} x an array
+   * @param {*} v value
+   * @param {function?} fn map function (v, i, x)
+   * @param {object?} ths this argument
+   * @returns {number} occurrences
+   */
+  export function countOn<T>(x: T[], v: T, fn?: (v: T, i: number, x: T[]) => T[], ths?: object): number;
+
+  /**
+   * Breaks array at given indices.
+   * @param {Array} x an array
+   * @param {Array<number>} is split indices (sorted)
+   * @returns {Array<Array>} [...pieces]
+   */
+  export function cut<T>(x: T[], is: number[]): T[][];
+
+  /**
+   * Breaks array after given indices.
+   * @param {Array} x an array
+   * @param {Array<number>} is split indices (sorted)
+   * @returns {Array<Array>} [...pieces]
+   */
+  export function cutRight<T>(x: T[], is: number[]): T[][];
+
+  /**
+   * Gives values of an array not present in another.
+   * @param {Array} x an array
+   * @param {Array} y another array
+   * @param {function?} fn compare function (a, b)
+   * @returns {Array}
+   */
+  export function difference<T>(x: T[], y: T[], fn?: (a: T, b: T) => boolean): T[];
+
+  /**
+   * Gives values of an array not present in another.
+   * @param {Array} x an array
+   * @param {Array} y another array
+   * @param {function?} fn map function (v, i, x)
+   * @param {object?} ths this argument
+   * @returns {Array}
+   */
+  export function differenceOn<T>(x: T[], y: T[], fn?: (v: T, i: number, x: T[]) => T[], ths?: object): T[];
+
+  /**
+   * Fills with given value.
+   * @param {Array} x an array
+   * @param {*} v value
+   * @param {number?} i start index (0)
+   * @param {number?} I end index (x.length-1)
+   * @returns {Array} filled
+   */
+  export function fill<T>(x: T[], v: T, i?: number, I?: number): T[];
+
+  /**
+   * Keeps the values which pass the test.
+   * @param {Array} x an array (updated)
+   * @param {function} fn filter function (v, i, x)
+   * @param {object?} ths this argument
+   * @returns {Array} x
+   */
+  export function filter$<T>(x: T[], fn: (v: T, i: number, x?: T[]) => T[], ths?: object): T[];
+
+  /**
+   * Finds indices of values passing the test.
+   * @param {Array} x an array
+   * @param {function} fn test function (v, i, x)
+   * @param {object?} ths this argument
+   * @returns {Array<number>} [...indices]
+   */
+  export function findIndices<T>(x: T[], fn: (v: T, i?: number, x?: T[]) => boolean, ths?: object): number[];
+  /**
+   * Finds index of rightmost value passing the test.
+   * @param {Array} x an array
+   * @param {function} fn test function (v, i, x)
+   * @param {object?} ths this argument
+   * @returns {number} index of value, -1 if not found
+   */
+  export function findRight<T>(x: T[], fn: (v: T, i?: number, x?: T[]) => boolean, ths?: object): number;
+
+  /**
+   * Flattens nested array to given depth.
+   * @param {Array} x a nested array
+   * @param {number?} dep maximum depth (-1)
+   * @returns {Array} flattened
+   */
+  export function flatten<T>(x: T[], dep?: number): T[];
```

Comment:
The types of `x` and the `return` are not same here.

```javascript
var x = [[1, 2], [3, [4, [5]]]];   // type (Array[])
array.flatten(x);
// [1, 2, 3, 4, 5]      (type number[])
```

Maybe this would do:

```suggestion
  export function flatten<T,U>(x: T[], dep?: number): U[];
```

Reply:
Ah I see, thanks for the heads up.